### PR TITLE
HTTP2: Ensure we never send EndStream after RST_STREAM

### LINKED ISSUE #39586: HTTP2: Hang when sending multiple requests from a HttpClient
I'm not sure what the cause of this hang is, it appears to happen at HttpClient.SendAsync when the client has made at least one previous call.

Repro:
1. `git clone https://github.com/JamesNK/grpc-dotnet.git`
2. `git checkout jamesnk/hang-multiple-calls`
3. `dotnet test test\FunctionalTests --filter Name~DuplexStream_SendToUnimplementedMethodAfterResponseReceived_Hang`

Logic in the test is like:
1. Client makes HTTP/2 duplex call to the server
2. Server immediately returns 200 and ends request.
3. Client inspects headers.
4. Client completes request stream.
5. Go to 1.

It consistently hangs making the second or third call.

```
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100-preview8-013248
 Commit:    135af55c46
```

### LINKED ISSUE #39819: HTTP2: Race between response body Dispose and request body completion
Found during GRPC stress.

If the client finishes sending the request body at approximately the same time that it disposes the response body without reading to the end, then there's a race between sending EndStream for the request body and RST_STREAM caused by the dispose. If the RST_STREAM ends up getting sent first, then this is a protocol violation.

### PR DESCRIPTION
**Pull request by geoffkizer**
Fixes #39586 
Fixes #39819 

There's a timing gap currently which can result in us sending EndStream on the request body after we've already sent RST_STREAM, which is not allowed.

Fix this by ensuring we never send EndStream if we've already sent RST_STREAM.

@danmosemsft This will need to get ported to the 3.0 branch -- GRPC stress issue.

@eiriktsarpalis Can we add a stress variation that tries to induce this issue? See description in #39819 (edit: fixed issue # here)

### CODE REVIEW

**Comment by geoffkizer** [comment]
I'm running our HttpStress on this now.

I ran GRPC stress on this as well, and did not see the original issue described in #39819 repro. However, I had to do this with a hacked build to avoid https://github.com/aspnet/AspNetCore/issues/12628.

**Comment by geoffkizer** [comment]
HttpStress has been running for an hour, no failures.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs [hunk 1]
```diff
@@ -232,28 +244,34 @@ public async Task<bool> WaitFor100ContinueAsync(CancellationToken cancellationTo
...
+                Debug.Assert(_responseCompletionState != StreamCompletionState.InProgress);
+                Debug.Assert(_requestCompletionState == StreamCompletionState.Failed || _responseCompletionState == StreamCompletionState.Failed,
+                    "Reset called but neither request nor response is failed");
+
+                if (NetEventSource.IsEnabled) Trace($"Stream reset. This is a test. Request={_requestCompletionState}, Response={_responseCompletionState}.");
+                // Don't send a RST_STREAM if we've already received one from the server.
+                if (_resetException == null)
                 {
-                    return;
+                    _connection.LogExceptions(_connection.SendRstStreamAsync(_streamId, Http2ProtocolErrorCode.Cancel));
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: Is Http2ProtocolErrorCode.Cancel always the right error code to send?  It looks like we might send this if we experienced some kind of error; should that be represented to the server as Cancel?  I'm fine if the answer is "yes", just want to make sure.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs [hunk 2]
```diff
@@ -232,28 +244,34 @@ public async Task<bool> WaitFor100ContinueAsync(CancellationToken cancellationTo
...
-
-                if (_requestCompletionState == StreamCompletionState.InProgress || _responseCompletionState == StreamCompletionState.InProgress)
+                Debug.Assert(_requestCompletionState != StreamCompletionState.InProgress);
+                Debug.Assert(_responseCompletionState != StreamCompletionState.InProgress);
+                Debug.Assert(_requestCompletionState == StreamCompletionState.Failed || _responseCompletionState == StreamCompletionState.Failed,
+                    "Reset called but neither request nor response is failed");
+
+                if (NetEventSource.IsEnabled) Trace($"Stream reset. This is a test. Request={_requestCompletionState}, Response={_responseCompletionState}.");
+                // Don't send a RST_STREAM if we've already received one from the server.
+                if (_resetException == null)
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: From a quick look at the source, I believe the answer is "yes", but we only ever set this while holding the same SyncObject lock, correct?  Just want to validate there's no chance we could see it's null, enter the if block, only to then have another thread set the exception.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs [hunk 3]
```diff
@@ -232,28 +244,34 @@ public async Task<bool> WaitFor100ContinueAsync(CancellationToken cancellationTo
...
             {
                 Debug.Assert(Monitor.IsEntered(SyncObject));
-                Debug.Assert(_requestCompletionState != StreamCompletionState.InProgress || _responseCompletionState != StreamCompletionState.InProgress,
-                    $"CheckForCompletion called but neither request nor response is completed");
-
-                if (_requestCompletionState == StreamCompletionState.InProgress || _responseCompletionState == StreamCompletionState.InProgress)
+                Debug.Assert(_requestCompletionState != StreamCompletionState.InProgress);
+                Debug.Assert(_responseCompletionState != StreamCompletionState.InProgress);
+                Debug.Assert(_requestCompletionState == StreamCompletionState.Failed || _responseCompletionState == StreamCompletionState.Failed,
+                    "Reset called but neither request nor response is failed");
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: Just to confirm, another way of stating these three asserts is:
```C#
Debug.Assert(
    (_requestCompletionState == StreamCompletionState.Failed && _responseCompletionState == StreamCompletionState.Completed) ||
    (_requestCompletionState == StreamCompletionState.Completed && _responseCompletionState == StreamCompletionState.Failed));
```
right?

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: Oh, no, I guess they could both be failed.  Ok.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs (hunk 1 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: https://tools.ietf.org/html/rfc7540#section-7

The only other error that I can see might apply here is INTERNAL_ERROR.

But honestly, I don't think it matters much.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs (hunk 2 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: There's only once place this is set: during OnReset, and the SyncObject lock is held there.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs (hunk 3 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: Yes and yes. Both must be Completed or Failed. At least one must be Failed.

**Comment by danmosemsft** [comment]
Let's get it into 3.0 then. 

Ask mode starts Monday.

**Comment by JamesNK** [comment]
Ping me when this is in 3.0 and I'll update gRPC to the latest SDK and re-test https://github.com/dotnet/corefx/issues/39586. I can consistently repro it.

**Comment by geoffkizer** [comment]
It's in master now, will be ported to 3.0 soon (today, hopefully). https://github.com/dotnet/corefx/pull/39874

**Comment by karelz** [comment]
This PR went into master - fixing milestone to 5.0.
It will be ported into release/3.0 branch in PR #39874.

**Pull request merged**